File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        
        vector<vector<int>> ans;
        
        for(int i=0; i<nums.size(); i++){
            int li = i+1;
            int ri = nums.size()-1;
            int sum = 0 - nums[i];
            
            if(i == 0 || (i > 0 && nums[i] != nums[i-1])){
            
            while(li < ri){
                if(nums[li] + nums[ri] == sum){
                    vector<int> temp;
                    temp.push_back(nums[i]);
                    temp.push_back(nums[li]);
                    temp.push_back(nums[ri]);
                    
                    ans.push_back(temp);
                    
                    while(li < ri && nums[li+1] == nums[li])li++;
                    while(li < ri && nums[ri-1] == nums[ri])ri--;
                    li++;
                    ri--;
                }else if(nums[li] + nums[ri] < sum) li++;
                else ri--;
            }
            }
            //while(i < nums.size() && nums[i] == nums[i++])i++;
        }
        
        
        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        
#         -4, -1, -1, 0, 1, 2
        
        for i in range(0, len(nums)-2):
            l, r = i+1, len(nums)-1
            
            if(nums[i] > 0):
                break
                
            if i>0 and nums[i]==nums[i-1]:
                continue
            
            target = 0 - nums[i]
            while l<r:
                s = nums[l] + nums[r]
                if s==target:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while nums[l]==nums[l-1] and l<r:
                        l+=1
                elif s < target:
                    l += 1
                else:
                    r -= 1
        
        return res
            
            
        